semantics/executable-spec/src/Data/Coders.hs

```diff
@@ -155,162 +190,228 @@ wrapCBORArray len contents =
 
 -- ========================================================
+-- Subsidary classes and datatype used in the Coders scheme
+-- =========================================================
 
 -- | Some CBOR instances wrap encoding sequences with prefixes and suffixes. I.e.
---  prefix , encode, encode, encode , ... , suffix
---  A type that MUST do this is called 'Open. Other types are called 'Closed.
---  The point is that a Closed type may have a prefix, but it can still be 'inlined'
---  by using the context it appears in, inside another Closed type. It does so by
---  sharing the prefix of its containing type, and using positional context.
---  In terms of Open and Closed, a datatype with more than one constructor is Open,
---  and one with only one constructor is Closed (we call these records). Primitive
---  types with no constructors can also be inlined, so we mark them as closed.
-data Wrapped = Open | Closed
+--  prefix , encode, encode, encode , ... , suffix.
+--  There are two kinds of wrapping coders: Nary sums, and Sparsely encoded products.
+--  Coders in these classes can only be used when they are wrapped by their
+--  elimination forms Summand and SparseKeyed
+
+data Wrapped = Nary | Thin | Closed
+  deriving Typeable
 
+-- | To ensure that wrapped coders are safely closed, we use this class.
+-- Each wrapped constructor can only be applied to the right kind of args.
+class Applys (con::Wrapped)  (arg::Wrapped) | con -> arg where
+
+-- | Analogous to paired ToCBOR and FromCBOR instances with out freezing out
+--   alternate ways to code. There can be multiple Duals with the same type.
+data Dual t = Dual (t -> Encoding) (forall s . Decoder s t)
+
+-- | A Field pairs an update function and a decoder for one field of a Sparse record.
+data Field t where
+  Field:: (x -> t -> t) -> Decode 'Closed x -> Field t
+
+-- ===========================================================
+-- The coders and the decoders as GADT datatypes
 -- ===========================================================
 
 data Encode (w :: Wrapped) t where
-  Sum :: t -> Word -> Encode 'Open t
-  Rec :: t -> Encode 'Closed t
+  Rec :: t -> Encode 'Closed t         -- No closer needed, already closed
+  Sum :: t -> Word -> Encode 'Nary t   -- Its closer is Summands
+  Sparse :: t -> Encode 'Thin t        -- Its closer is SparseKeyed
   To :: ToCBOR a => a -> Encode 'Closed a
   E :: (t -> Encoding) -> t -> Encode 'Closed t
-  ApplyE :: Encode w (a -> t) -> Encode 'Closed a -> Encode w t
-
--- The Wrapped index of ApplyE is determined by the index
--- at the bottom of its left spine. The LEFT arg of ApplyE
--- must be a function type, and the only Encode with function
--- types are (Sum c tag) and (Rec c). So if the leftmost spine
--- is (Sum c tag) it is 'Open, and if is (Rec c) it is 'Closed.
--- The RIGHT arg of ApplyE must be 'Closed. This allows us to
--- inline anything in a RIGHT arg, supporting CBORGroup capability.
+  ED :: Dual t -> t -> Encode 'Closed t
```

Comment:
I agree,it could be a function
ed (Dual enc dec) x = E enc x
But that obscures the nice visual inspection of Duality, between (ED dual t) and (DD dual)
so lets leave it for now.